substrate/frame/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,340 @@
+	Positive(Balance),
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use crate::*;
+	use frame_election_provider_support::VoteWeight;
+	use frame_support::pallet_prelude::*;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Currency: FnInspect<Self::AccountId, Balance = BalanceOf<Self>>;
+
+		/// The staking interface.
+		type Staking: StakingInterface<AccountId = Self::AccountId>;
+
+		/// Something that provides a *best-effort* sorted list of voters.
+		///
+		/// To keep the load off the chain as much as possible, changes made to the staked amount
+		/// via rewards and slashes are dropped and thus need to be manually updated through
+		/// extrinsics. In case of `bags-list`, this always means using `rebag` and `putInFrontOf`.
+		type VoterList: SortedListProvider<Self::AccountId, Score = VoteWeight>;
+
+		/// Something that provides an *always* sorted list of targets.
+		type TargetList: SortedListProvider<Self::AccountId, Score = VoteWeight>;
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Returns the vote weight of a staker based on its current *active* stake, as returned by
+		/// the staking interface.
+		pub(crate) fn active_vote_of(who: &T::AccountId) -> VoteWeight {
+			T::Staking::stake(who)
+				.map(|s| Self::to_vote(s.active))
+				.defensive_unwrap_or_default()
+		}
+
+		/// Converts a staker's balance to its vote weight.
+		pub(crate) fn to_vote(balance: BalanceOf<T>) -> VoteWeight {
+			<T::Staking as StakingInterface>::CurrencyToVote::to_vote(
+				balance,
+				T::Currency::total_issuance(),
+			)
+		}
+
+		/// Updates a staker's score by increasing/decreasing an imbalance of the current score in
+		/// the list.
+		pub(crate) fn update_score<L>(who: &T::AccountId, imbalance: StakeImbalance<VoteWeight>)
+		where
+			L: SortedListProvider<AccountIdOf<T>, Score = VoteWeight>,
+		{
+			// there may be nominators who nominate a non-existant validator. if that's the case,
+			// move on.
+			if !L::contains(who) {
+				defensive!("`update_score` on non-existant staker {}", who);
+				return
+			}
+
+			match imbalance {
+				StakeImbalance::Positive(imbalance) => {
+					let _ = L::on_increase(who, imbalance).defensive_proof(
+						"staker should exist in the list, otherwise returned earlier.",
+					);
+				},
+				StakeImbalance::Negative(imbalance) => {
+					let current_score = L::get_score(who)
+						.expect("staker exists in the list as per the check above; qed.");
+
+					// if decreasing the imbalance makes the score lower than 0, the node will be
+					// removed from the list when calling `L::on_decrease`, which is not expected.
+					// Instead, we call `L::on_update` to set the score as 0. The node will be
+					// removed when `on_*_removed` is called.
```

Comment:
Should score ever go below zero?

Reply:
No, currently we use `current_score.defensive_saturating_sub(imbalance)`.